Support one whereHas for several filters.

IndexZer0 opened the pull request:
This PR adds `AllowedRelationshipFilter` class to be used when wanting to group multiple `AllowedFilter`'s into the same exists query.

### Problem:
When using multiple `AllowedFilter`'s for nested relations - the package adds mutliple exist clauses to the query.

```php
->allowedFilters([
    AllowedFilter::exact('relatedModels.id'),
    AllowedFilter::exact('relatedModels.name'),
    AllowedFilter::exact('relatedModels.nestedRelatedModels.id'),
    AllowedFilter::exact('relatedModels.nestedRelatedModels.name'),
])
```

```sql
SELECT
    *
FROM
    `test_models`
WHERE
    EXISTS (
        SELECT
            *
        FROM
            `related_models`
        WHERE
            `test_models`.`id` = `related_models`.`test_model_id`
            AND `related_models`.`id` = 1)
        AND EXISTS (
            SELECT
                *
            FROM
                `related_models`
            WHERE
                `test_models`.`id` = `related_models`.`test_model_id`
                AND `related_models`.`name` = 'asdf')
            AND EXISTS (
                SELECT
                    *
                FROM
                    `related_models`
                WHERE
                    `test_models`.`id` = `related_models`.`test_model_id`
                    AND EXISTS (
                        SELECT
                            *
                        FROM
                            `nested_related_models`
                        WHERE
                            `related_models`.`id` = `nested_related_models`.`related_model_id`
                            AND `nested_related_models`.`id` = 1))
                    AND EXISTS (
                        SELECT
                            *
                        FROM
                            `related_models`
                        WHERE
                            `test_models`.`id` = `related_models`.`test_model_id`
                            AND EXISTS (
                                SELECT
                                    *
                                FROM
                                    `nested_related_models`
                                WHERE
                                    `related_models`.`id` = `nested_related_models`.`related_model_id`
                                    AND `nested_related_models`.`name` = 'ghjk'))
```

Resulting in 6 `EXISTS` queries.

### Desired Query

```sql
SELECT
    *
FROM
    `test_models`
WHERE
    EXISTS (
        SELECT
            *
        FROM
            `related_models`
        WHERE
            `test_models`.`id` = `related_models`.`test_model_id`
            AND `related_models`.`id` = 1
            AND `related_models`.`name` = 'asdf'
            AND EXISTS (
                SELECT
                    *
                FROM
                    `nested_related_models`
                WHERE
                    `related_models`.`id` = `nested_related_models`.`related_model_id`
                    AND `nested_related_models`.`id` = 1
                    AND `nested_related_models`.`name` = 'ghjk'))
```

Resulting in 2 `EXISTS` queries.

I have seen various requests for this in the past, for example

https://github.com/spatie/laravel-query-builder/discussions/634

https://github.com/spatie/laravel-query-builder/discussions/663

---

At this point just looking for your thoughts and feedback on this. I'm not particually happy with the solution in this PR, feels quite hacky around the existing `AllowedFilter` class.

To me, this feels like quite an important missing feature of this package and if the core maintainers could give some feedback, insight and pointers on this then I'm happy to make the changes and then update documentation to match.

AlexVanderbist commented on src/AllowedRelationshipFilter.php:
```diff
@@ -0,0 +1,73 @@
+            $this->allowedFilters->each(
+                function (AllowedFilterContract $allowedFilter) use ($query, $value) {
+                    $allowedFilter->filter(
+                        QueryBuilder::for($query),
```
I might be missing something, but why not pass the current `$query` object? It should already be an instance of `QueryBuilder`

IndexZer0 replied:
Tried to swap to `$query` and get error.

`Spatie\QueryBuilder\AllowedFilter::filter(): Argument #1 ($query) must be of type Spatie\QueryBuilder\QueryBuilder, Illuminate\Database\Eloquent\Builder given, called in /src/AllowedRelationshipFilter.php on line 27`

Keeping this as is.

AlexVanderbist commented on src/AllowedFilter.php:
```diff
@@ -199,4 +200,19 @@ protected function resolveValueForFiltering($value)
         return ! $this->ignored->contains($value) ? $value : null;
     }
+
+    public function isRequested(QueryBuilderRequest $request): bool
+    {
+        return $request->filters()->has($this->getName());
+    }
+
+    public function getValueFromRequest(QueryBuilderRequest $request): mixed
+    {
+        return $request->filters()->get($this->getName());
+    }
```
Nice little refactor 👍

AlexVanderbist commented:
Hey @IndexZer0, thanks for the extensive research and PR! As the impact on the existing filters is minimal, I'm happy to merge this in its current state in the v6 release and see what feedback comes in once people start using the `AllowedRelationshipFilter::group()` method.

Is there anything specific you would like feedback on?

IndexZer0 commented:
> Hey @IndexZer0, thanks for the extensive research and PR! As the impact on the existing filters is minimal, I'm happy to merge this in its current state in the v6 release and see what feedback comes in once people start using the `AllowedRelationshipFilter::group()` method.
> 
> Is there anything specific you would like feedback on?

Thanks for reviewing @AlexVanderbist 

Not particulary after specific feedback if you're happy to merge this feature. Just was open to changing the implementation approach if you had any better ideas on how to structure this.

- Updated to this fork to current spatie:main commit.
- Fixed merge conflicts.
- Added brief example to documentation.
- My github actions show tests are failing.
  - Though these tests are failing in spatie:main too.
  - The newly added tests in this PR are passing currently.

FYI - As it turns out, this package is not going to suit my use case anymore so doubt I'd have the desire to update this PR again.

Hope you can get this merged soon before more merge conflicts :)
